### PR TITLE
Add libcxi and its dependencies

### PR DESCRIPTION
**Pull request by rbberger**
Add the `libcxi` library, which as been recently open-sourced by HPE/Cray.

### CODE REVIEW

**Comment by rbberger** [comment]
@hppritcha I'd be curious if we could get OpenMPI / libfabric built with this.

**Comment by hppritcha** [comment]
I will give it a try on one of our ss11 systems

File: var/spack/repos/builtin/packages/libcxi/package.py
Comment (haampie) [inline var/spack/repos/builtin/packages/libcxi/package.py]:
```suggestion
    depends_on("libfuse@2.9.7:2")  # configure fails with newer, bug?
```

happens sometimes due to API changes. did you file a bug?

File: var/spack/repos/builtin/packages/libcxi/package.py
Comment (haampie) [inline var/spack/repos/builtin/packages/libcxi/package.py:47]: did you open an issue for this hard-coded path?

**Comment by hppritcha** [comment]
silly question but what install args should I use for libfabric to make sure i get libfabric + cxi provider?
I was able apparently to build libcxi on one of our ss11 systems.

File: var/spack/repos/builtin/packages/libcxi/package.py
Comment (rbberger) [reply var/spack/repos/builtin/packages/libcxi/package.py]: https://github.com/HewlettPackard/shs-libcxi/issues/3

File: var/spack/repos/builtin/packages/libcxi/package.py
Comment (rbberger) [reply var/spack/repos/builtin/packages/libcxi/package.py:47]: https://github.com/HewlettPackard/shs-libcxi/issues/4

**Comment by rbberger** [comment]
@hppritcha :sweat_smile:  I was hoping you'd be faster at figuring that one out, but I can try myself later today.

**Comment by hppritcha** [comment]
hmmm i don't know about this PR.  What is its purpose?

```=> Error: ProcessError: Command exited with status 1:
    '/tmp/hpp/spack-stage/spack-stage-libfabric-1.22.0-gagbwp45bac2cy2nlowzdko7rhybckx2/spack-src/configure' '--prefix=/usr/projects/artab/users/hpp/spack/opt/spack/linux-sles15-sapphirerapids/gcc-12.2.0/libfabric-1.22.0-gagbwp45bac2cy2nlowzdko7rhybckx2' '--disable-debug' '--with-kdreg=no' '--enable-cxi=yes' '--enable-efa=no' '--enable-gni=no' '--enable-mlx=no' '--enable-mrail=no' '--enable-opx=no' '--enable-psm=no' '--enable-psm2=no' '--enable-psm3=no' '--enable-rxm=no' '--enable-rxd=no' '--enable-shm=no' '--enable-sockets=no' '--enable-tcp=no' '--enable-ucx=no' '--enable-udp=no' '--enable-usnic=no' '--enable-verbs=no' '--enable-xpmem=no'

1 error found in build log:
     210    checking json-c/json.h presence... yes
     211    checking for json-c/json.h... yes
     212    configure: looking for library without search path
     213    checking for json_object_get_type in -ljson-c... yes
     214    configure: cxi provider: disabled
     215    configure: WARNING: cxi provider was requested, but cannot be compiled
  >> 216    configure: error: Cannot continue
```
I hacked on libfabric spack to allow for selection of cxi fabrics variant and this is what resulted.

**Comment by rbberger** [comment]
Yes, that's as far as I got yesterday. Btw, here is the guix patch set: https://issues.guix.gnu.org/74419

**Comment by hppritcha** [comment]
I will poke a little bit later.  I would not approve this PR if it's only usable by vendor.

**Comment by rbberger** [comment]
@hppritcha pushed what is needed to have `libfabric fabrics=cxi` build.

**Comment by rbberger** [comment]
I was able to build OpenMPI 5 with this:
```yaml
spack:
  specs:
  - openmpi
  view: true
  concretizer:
    unify: true
  packages:
    openmpi:
      require: fabrics=ofi schedulers=slurm
    libfabric:
      require: fabrics=cxi,shm,sockets,tcp,udp
    curl:
      require: ~nghttp2
```

**Comment by hppritcha** [comment]
hmm still not building for me, now some problem with symbols with curl
```
collect2: error: ld returned 1 exit status
/usr/bin/ld: /usr/projects/artab/users/hpp/spack/opt/spack/linux-sles15-sapphirerapids/gcc-12.2.0/curl-8.10.1-2oyrt2ste52zzt2m2he7l5jrbhgzbbnx/lib/libcurl.so: undefined reference to `nghttp2_option_set_no_rfc9113_leading_and_trailing_ws_validation'
/usr/bin/ld: /usr/projects/artab/users/hpp/spack/opt/spack/linux-sles15-sapphirerapids/gcc-12.2.0/curl-8.10.1-2oyrt2ste52zzt2m2he7l5jrbhgzbbnx/lib/libcurl.so: undefined reference to `nghttp2_option_set_no_rfc9113_leading_and_trailing_ws_validation'
make[1]: *** [Makefile:12746: util/fi_strerror] Error 1
make[1]: *** Waiting for unfinished jobs....
collect2: error: ld returned 1 exit status
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:12740: util/fi_pingpong] Error 1
make[1]: *** [Makefile:12734: util/fi_info] Error 1
/usr/bin/ld: /usr/projects/artab/users/hpp/spack/opt/spack/linux-sles15-sapphirerapids/gcc-12.2.0/curl-8.10.1-2oyrt2ste52zzt2m2he7l5jrbhgzbbnx/lib/libcurl.so: undefined reference to `nghttp2_option_set_no_rfc9113_leading_and_trailing_ws_validation'
/usr/bin/ld: /usr/projects/artab/users/hpp/spack/opt/spack/linux-sles15-sapphirerapids/gcc-12.2.0/curl-8.10.1-2oyrt2ste52zzt2m2he7l5jrbhgzbbnx/lib/libcurl.so: undefined reference to `nghttp2_option_set_no_rfc9113_leading_and_trailing_ws_validation'
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:12648: prov/cxi/test/multinode/test_frmwk] Error 1
collect2: error: ld returned 1 exit status
/usr/bin/ld: /usr/projects/artab/users/hpp/spack/opt/spack/linux-sles15-sapphirerapids/gcc-12.2.0/curl-8.10.1-2oyrt2ste52zzt2m2he7l5jrbhgzbbnx/lib/libcurl.so: undefined reference to `nghttp2_option_set_no_rfc9113_leading_and_trailing_ws_validation'
make[1]: *** [Makefile:12638: prov/cxi/test/multinode/test_coll] Error 1
collect2: error: ld returned 1 exit status
/usr/bin/ld: /usr/projects/artab/users/hpp/spack/opt/spack/linux-sles15-sapphirerapids/gcc-12.2.0/curl-8.10.1-2oyrt2ste52zzt2m2he7l5jrbhgzbbnx/lib/libcurl.so: undefined reference to `nghttp2_option_set_no_rfc9113_leading_and_trailing_ws_validation'
make[1]: *** [Makefile:12658: prov/cxi/test/multinode/test_zbcoll] Error 1
collect2: error: ld returned 1 exit status
make[1]: *** [Makefile:12628: prov/cxi/test/multinode/test_barrier] Error 1
make[1]: Leaving directory '/tmp/hpp/spack-stage/spack-stage-libfabric-1.22.0-zaxwwosbc5d7mnocto2p7aqjdqxrqpbp/spack-src'
make: *** [Makefile:6816: all] Error 2
```

**Comment by hppritcha** [comment]
okay worked around that problem by install curl like this
```
spack install curl ~nghttp2
```

**Comment by hppritcha** [comment]
okay some basic smoke tests with Open MPI main branch using libfabric built using this libcxi + patches and a few hacks seems to work.  i will try this with a spack env next week.

**Comment by hppritcha** [comment]
i'm trying to build this thing now from a fresh spack check (well the libfabric with fabric=cxi) and still see problems when the libfabric make procedure gets to building binaries like fi_info 
```
 >> 774    //////////////////////////////////////////usr/projects/hpcsoft/tce/23-12/cos2-x86_64/linux-sles15-x86_64_v3/binutils/-/binutils-2.38-o7dkz5ywrg5s6sjbclt3i57af7p5zuq3/bin/ld: /usr/projects/artab/user
            s/hpp/spack/opt/spack/linux-sles15-sapphirerapids/gcc-12.3.0/curl-8.10.1-gqqdr2ucwtmfobnmzzqoozduoiumrrgp/lib/libcurl.so: undefined reference to `nghttp2_option_set_no_rfc9113_leading_and_trailing_w
            s_validation'
  >> 775    //////////////////////////////////////////usr/projects/hpcsoft/tce/23-12/cos2-x86_64/linux-sles15-x86_64_v3/binutils/-/binutils-2.38-o7dkz5ywrg5s6sjbclt3i57af7p5zuq3/bin/ld: /usr/projects/artab/user
            s/hpp/spack/opt/spack/linux-sles15-sapphirerapids/gcc-12.3.0/curl-8.10.1-gqqdr2ucwtmfobnmzzqoozduoiumrrgp/lib/libcurl.so: undefined reference to `nghttp2_option_set_no_rfc9113_leading_and_trailing_w
            s_validation'
  >> 776    collect2: error: ld returned 1 exit status
  >> 777    collect2: error: ld returned 1 exit status
  >> 778    make[1]: *** [Makefile:12583: util/fi_strerror] Error 1
     779    make[1]: *** Waiting for unfinished jobs....
  >> 780    make[1]: *** [Makefile:12577: util/fi_pingpong] Error 1
  >> 781    //////////////////////////////////////////usr/projects/hpcsoft/tce/23-12/cos2-x86_64/linux-sles15-x86_64_v3/binutils/-/binutils-2.38-o7dkz5ywrg5s6sjbclt3i57af7p5zuq3/bin/ld: /usr/projects/artab/user
            s/hpp/spack/opt/spack/linux-sles15-sapphirerapids/gcc-12.3.0/curl-8.10.1-gqqdr2ucwtmfobnmzzqoozduoiumrrgp/lib/libcurl.so: undefined reference to `nghttp2_option_set_no_rfc9113_leading_and_trailing_w
            s_validation'
  >> 782    collect2: error: ld returned 1 exit status
  >> 783    make[1]: *** [Makefile:12571: util/fi_info] Error 1
```
only way i've come up so far to fix this is to build curl with ~nghttp2.  Any ideas?  I've tried adding nghttp2 path to the LDFLAGS like with curl and json-c but to no avail.

**Comment by rbberger** [comment]
Must be related to this: https://github.com/curl/curl/issues/2516

**Comment by rbberger** [comment]
@hppritcha I added the necessary flags for `nghttp2` in libfabric.

**Comment by hppritcha** [comment]
okay this looks good to me.  i tested with IMB-MPI1 and various osu tests on one of our SS11 systems.  I think the libfabric piece needs to be split out into a separate PR?  but otherwise I approve.  thanks @rbberger this is a big step forward.

**Comment by rbberger** [comment]
@rajachan are you ok with these libfabric changes?

**Review by rajachan — APPROVED** [review]

**Comment by rbberger** [comment]
@haampie @alalazo I think we're good now. Unless you want me to split this into multiple PRs.

**Comment by etiennemlb** [comment]
@rbberger @hppritcha I had to change libfabric like so:
image

But then I ahve another issue, json-c is not linked via rpath for some reason, and causes issue when linking OpenMPI.

**Comment by rbberger** [comment]
@etiennemlb Try with my latest push.

**Comment by etiennemlb** [comment]
@rbberger thanks, ill try that tomorrow. Also, can a core dev tell me why in this case, json-c's lib directory is not in the rpath. The wrapper does not seem to be aware that it should add json-c's path Wl,-rpath.

**Comment by etiennemlb** [comment]
It builds. The performance of /mpi/pt2pt/osu_bw was lower than with mpich (83% of what mpich can provide). But at least it seems to work. (note I work on a SS11 system very similar to frontier or lumi). The tests were done on GPU nodes, although the osu bench only used cpu memory.

**Comment by biddisco** [comment]
I tried building, but had some issues which I fixed by explicitly passing paths to the configure. 
my branch changes are here https://github.com/biddisco/spack/tree/libcxi
No idea why I can't create a merge request into this branch. Thanks for the packages, very useful, will be testing the builds at CSCS.

**Comment by rbberger** [comment]
@biddisco thanks for those extra config options. That should make it more robust. I'll verify next week if we can now drop some of the changes in `setup_build_environment`.

**Comment by rbberger** [comment]
@etiennemlb Interesting. I've built it on Rocinante here at LANL, and I got near peak bandwidth (96%) with osu_bw. Maybe we should standardize our benchmark command-lines and compare the results. I still have to test the `+rocm` and `+cuda` combos. I'm planning to do `+cuda` on Venado, and `+rocm` on LLNL systems.

**Comment by etiennemlb** [comment]
@rbberger I agree about the need for standardization, I'll share my srun command and the osu spack spec.

I have tried using +rocm on openmpi and osu (that required some minor tweaks to make the thing build). In any case, I got error when doing device to device bandwidth measurement. AMD provides an OMPI+UCX GPU enabled recipe, but again, the performance was not great. 

Later, I'll also check the perf on MI300A nodes and on AMD EPYC GENOA nodes.

File: var/spack/repos/builtin/packages/libfabric/package.py
Comment (haampie) [inline var/spack/repos/builtin/packages/libfabric/package.py]: why build?

File: var/spack/repos/builtin/packages/libfabric/package.py
Comment (haampie) [inline var/spack/repos/builtin/packages/libfabric/package.py]: This shouldn't be here. Looks like in @hppritcha's case curl was linked dynamically, and the linker tried and failed to resolve nghttp2 symbols from transitive deps. That's a curl problem, not a libfabric issue.

Comment (haampie) [reply var/spack/repos/builtin/packages/libfabric/package.py]: Also @hppritcha do you still see this issue where you cannot link to `libcurl.so`? If so can you share the spec and `ldd libcurl.so`?

I cannot reproduce. `ldd curl.so` resolves `libnghttp2.so.14` for me, and it has the symbol `nghttp2_option_set_no_rfc9113_leading_and_trailing_ws_validation`.

Looks like that symbol was introduced in v1.50, so I can only see this could happen if nghttp2 was external, and then downgraded (?). Or if you removed the spack prefix of nghttp2, and it falls back to an old system version. Both sound unlikely.

File: var/spack/repos/builtin/packages/libfabric/package.py
Comment (rbberger) [reply var/spack/repos/builtin/packages/libfabric/package.py]: yeah, right... 🤦‍♂️

File: var/spack/repos/builtin/packages/libfabric/package.py
Comment (rbberger) [reply var/spack/repos/builtin/packages/libfabric/package.py]: I ran into the same issue. `curl` wasn't external, but I need to retest now that we have proper configure flags passed to libfabric.

Comment (rbberger) [reply var/spack/repos/builtin/packages/libfabric/package.py]: At least on my end, the explicit configure options, removing the manual LDFLAGS and changing the dependency type has now resolved the issue. @hppritcha could you check it again with the current state?
@etiennemlb for me it looks like that `libfabric.so` now has the correct rpaths even without having to manually add them.

File: var/spack/repos/builtin/packages/libfabric/package.py
Comment (rbberger) [reply var/spack/repos/builtin/packages/libfabric/package.py]: Fixed.

File: var/spack/repos/builtin/packages/libfabric/package.py
Comment (hppritcha) [reply var/spack/repos/builtin/packages/libfabric/package.py]: i'll try to get to this tomorrow - 12/10/24

**Comment by hppritcha** [comment]
this PR works without any special workarounds for me now on my ss11 system.

**Comment by rbberger** [comment]
resolved the merge conflict. @haampie I believe this is good to go.

**Comment by etiennemlb** [comment]
> resolved the merge conflict. @haampie I believe this is good to go.

LGTM too but could you share you benchmarking setup. Spack specs, bindings, hardware used, etc..

**Comment by rbberger** [comment]
> > resolved the merge conflict. @haampie I believe this is good to go.
> 
> LGTM too but could you share you benchmarking setup. Spack specs, bindings, hardware used, etc..

@etiennemlb Nothing fancy, just something like the following plus cluster specific compiler and externals specs:
```yaml
spack:
  specs:
  - openmpi
  - osu-micro-benchmarks
  view: true
  concretizer:
    unify: true
  packages:
    openmpi:
      require: fabrics=ofi schedulers=slurm
    libfabric:
      require: fabrics=cxi,shm,sockets,tcp,udp
```

```yaml
spack:
  specs:
  - mpich
  - osu-micro-benchmarks
  view: true
  concretizer:
    unify: true
  packages:
    mpi:
      require: mpich
    mpich:
      require: netmod=ofi +slurm
    libfabric:
      require: fabrics=cxi,shm,sockets,tcp,udp
```

I'm measuring the the host to host bandwidth by launching via `srun -N2 -n 2 --mpi=pmix osu_get_bw`

My main observation at the moment is that for sizes 256 - 16384 CrayMPICH seems to be outperforming both OpenMPI and MPICH, which likely can be attributed to tuning, while the peak bandwidth is best for OpenMPI for the largest sizes. I haven't had the time to do CUDA and ROCm device to device benchmarks. But this shouldn't hold back this PR.

File: var/spack/repos/builtin/packages/libcxi/package.py
Comment (haampie) [inline var/spack/repos/builtin/packages/libcxi/package.py]:
```suggestion
    variant("level_zero", default=False, description="Build with OneAPI Level Zero support")
```

this is the common name of the variant since last week #47985

Comment (rbberger) [reply var/spack/repos/builtin/packages/libcxi/package.py]: Fixed.

**Review by haampie — APPROVED** [review]

**Comment by green-br** [comment]
Very interesting change and something I have been thinking about.  Can I check whether anyone has looked into https://github.com/HewlettPackard/shs-libfabric since I got the impression there was still changes to upstream to libfabric?

**Pull request merged**
